# HG changeset patch
# User Matti Picus <matti.picus@gmail.com>
# Date 1691873417 -10800
#      Sat Aug 12 23:50:17 2023 +0300
# Branch conda-3.9-v7.3.12
# Node ID 7a1d0ec8c2b0c1dba4cfa57fb98065e960321d23
# Parent  8e51bb8ff5066cfcc145c74aaaaf012c39dbc510
#3978: report exceptions that are created when forcing an oefmt as unraisable

diff -r 8e51bb8ff506 -r 7a1d0ec8c2b0 pypy/interpreter/error.py
--- a/pypy/interpreter/error.py	Wed Jul 19 10:57:25 2023 +0200
+++ b/pypy/interpreter/error.py	Sat Aug 12 23:50:17 2023 +0300
@@ -569,13 +569,23 @@
                         result = str(value)
                         lgt += len(result)
                     elif fmt == 'R':
-                        s = space.repr(value)
-                        result = space.utf8_w(s)
-                        lgt += space.len_w(s)
+                        try:
+                            w_s = space.repr(value)
+                            result = space.utf8_w(w_s)
+                            lgt += space.len_w(w_s)
+                        except OperationError as e:
+                            e.write_unraisable(space, "exception formatting: ", value)
+                            result = '<repr raised>'
+                            lgt += len(result)
                     elif fmt == 'S':
-                        s = space.str(value)
-                        result = space.utf8_w(s)
-                        lgt += space.len_w(s)
+                        try:
+                            s = space.str(value)
+                            result = space.utf8_w(s)
+                            lgt += space.len_w(s)
+                        except OperationError as e:
+                            e.write_unraisable(space, "exception formatting: ", value)
+                            result = '<str raised>'
+                            lgt += len(result)
                     elif fmt == 'T':
                         result = space.type(value).name
                         lgt += rutf8.codepoints_in_utf8(result)
